src/gc.rs

```diff
@@ -0,0 +1,257 @@
+    num::NonZeroUsize,
+    ops::{CoerceUnsized, Deref, DerefMut},
+    ptr::NonNull,
+};
+
+use crate::boehm;
+use crate::GC_ALLOCATOR;
+
+/// A garbage collected pointer. 'Gc' stands for 'Garbage collected'.
+///
+/// The type `Gc<T>` provides shared ownership of a value of type `T`,
+/// allocted in the heap. `Gc` pointers are `Copyable`, so new pointers to
+/// the same value in the heap can be produced trivially. The lifetime of
+/// `T` is tracked automatically: it is freed when the application
+/// determines that no references to `T` are in scope. This does not happen
+/// deterministically, and no guarantees are given about when a value
+/// managed by `Gc` is freed.
+///
+/// Shared references in Rust disallow mutation by default, and `Gc` is no
+/// exception: you cannot generally obtain a mutable reference to something
+/// inside an `Gc`. If you need mutability, put a `Cell` or `RefCell` inside
+/// the `Gc`.
+///
+/// Unlike `Rc<T>`, cycles between `Gc` pointers are allowed and can be
+/// deallocated without issue.
+///
+/// `Gc<T>` automatically dereferences to `T` (via the `Deref` trait), so
+/// you can call `T`'s methods on a value of type `Gc<T>`.
+///
+/// `Gc<T>` is implemented using a tracing mark-sweep garbage collection
+/// algorithm. This means that by using `Gc` pointers in a Rust application,
+/// you pull in the overhead of a run-time garbage collector to manage and
+/// free `Gc` values behind the scenes.
+#[derive(PartialEq, Eq, Debug)]
+pub struct Gc<T: ?Sized> {
+    ptr: NonNull<GcBox<T>>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Gc<T> {
+    /// Constructs a new `Gc<T>`.
+    pub fn new(v: T) -> Self {
+        Gc {
+            ptr: unsafe { NonNull::new_unchecked(GcBox::new(v)) },
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Constructs a new `Gc<MaybeUninit<T>>` which is capable of storing data
+    /// up-to the size permissible by `layout`.
+    ///
+    /// This can be useful if you want to store a value with a custom layout,
+    /// but have the collector treat the value as if it were T.
+    ///
+    /// `layout` must be at least as large as `T`, and have an alignment which
+    /// is the same, or bigger than `T`.
+    pub fn new_from_layout(layout: Layout) -> Option<Gc<MaybeUninit<T>>> {
+        let tl = Layout::new::<T>();
+        if layout.size() < tl.size() && layout.align() >= tl.align() {
+            return None;
+        }
+        Some(Gc::from_inner(GcBox::new_from_layout(layout)))
+    }
+}
+
+impl Gc<dyn Any> {
+    pub fn downcast<T: Any>(&self) -> Result<Gc<T>, Gc<dyn Any>> {
+        if (*self).is::<T>() {
+            let ptr = self.ptr.cast::<GcBox<T>>();
+            forget(self);
+            Ok(Gc::from_inner(ptr))
+        } else {
+            Err(Gc::from_inner(self.ptr))
+        }
+    }
+}
+
+impl<T: ?Sized> Gc<T> {
+    /// Get a raw pointer to the underlying value `T`.
+    pub fn into_raw(this: Self) -> *const T {
```

Comment:
Is there any reason why this has to consume the `Gc`? [I think we've had this discussion before, but I no longer remember what the reason was!]

Reply:
This was for API compatibility with `Rc`. Having this signature made using `RustPython` with GC a lot easier.

Reply:
Fair enough.